cvmfs/network/jobinfo.cc

```diff
@@ -80,4 +88,211 @@ void JobInfo::Init() {
+ * otherwise return false and leave the input character in the first
+ * output character.
+ */
+bool JobInfo::EscapeUrlChar(unsigned char input,
+                            char output[3]) {
+  if (((input >= '0') && (input <= '9')) ||
+      ((input >= 'A') && (input <= 'Z')) ||
+      ((input >= 'a') && (input <= 'z')) ||
+      (input == '/') || (input == ':') || (input == '.') ||
+      (input == '@') ||
+      (input == '+') || (input == '-') ||
+      (input == '_') || (input == '~') ||
+      (input == '[') || (input == ']') || (input == ','))
+  {
+    output[0] = static_cast<char>(input);
+    return false;
+  }
+
+  output[0] = '%';
+  output[1] = static_cast<char>(
+                             (input / 16) + ((input / 16 <= 9) ? '0' : 'A'-10));
+  output[2] = static_cast<char>(
+                             (input % 16) + ((input % 16 <= 9) ? '0' : 'A'-10));
+  return true;
+}
+
+
+static std::string EscapeHeader(const std::string &header)
+{
+  std::string escaped = "";
+  char escaped_char[3];
+  for (char c : header) {
+    if (JobInfo::EscapeUrlChar(c, escaped_char)) {
+      for (unsigned j = 0; j < 3; ++j) {
+        escaped += escaped_char[j];
+      }
+    } else {
+      escaped += escaped_char[0];
+    }
+  }
+
+  return escaped;
+}
+
+
+/*
+ * Return the filled-in template of CVMFS_INFO_HEADER
+ */
+
+const std::string JobInfo::info_header_contents(const std::string templ) {
+  std::string answer = "";
+  std::string key;
+  int parsemode = 0;  // 0 default, 1 after '%', 2 after '%{'
+  int envsize = -1;
+  char *envbuf = NULL;
+
+  for (char c : templ) {
+    switch (parsemode) {
+    case 0:
+      if (c == '%') {
+        parsemode = 1;
+        key = "";
+      } else {
+        answer += c;
+      }
+      break;
+    case 1:
+      if (c == '{') {
+        parsemode = 2;
+      } else {
+        answer += '%';
+        answer += c;
+      }
+      break;
+    case 2:
+      if (c != '}') {
+        key += c;
+      } else {
+        parsemode = 0;
+        if (key == "path") {
+          if (path_info_ != NULL) {
+            answer += EscapeHeader(*path_info_);
+          } else {
+            answer += "-";
+          }
+        } else if (key == "pid") {
+          if (pid_ != (pid_t) -1) {
+            answer += StringifyInt(pid_);
+          } else {
+            answer += "-";
+          }
+        } else if (key == "uid") {
+          if (uid_ != (uid_t) -1) {
+            answer += StringifyInt(uid_);
+          } else {
+            answer += "-";
+          }
+        } else if (key == "gid") {
+          if (gid_ != (gid_t) -1) {
+            answer += StringifyInt(gid_);
+          } else {
+            answer += "-";
+          }
+        } else if (key.substr(0, 4) == "env:") {
+          if (envsize == -1) {
+            envsize = 0;
+            if (pid_ != (pid_t) -1) {
+              ObtainDacReadSearchCapability();
+              ObtainSysPtraceCapability();
```

Comment:
We would do this for each request... how heavy is this? Should we warn users when using `env:` in the template?

Reply:
Also, I think the regular and the external download manager may step on each others feet obtaining and dropping capabilities. Perhaps we need an helper process with elevated capabilities to read the environment, similar to the authorization helper.

Reply:
> We would do this for each request... how heavy is this? Should we warn users when using `env:` in the template?

Obtaining & dropping capabilities are a trivial amount of time.  The longer time is reading & parsing the environment, but it is all in memory and I assume it is negligible compared to the time of doing a download.  This doesn't happen when reading from cache.

Reply:
> Also, I think the regular and the external download manager may step on each others feet obtaining and dropping capabilities. Perhaps we need an helper process with elevated capabilities to read the environment, similar to the authorization helper.

Maybe, but that's a very heavyweight solution.  Perhaps there should be a lock, or perhaps it should obtain capabilities and not drop them.  On the other hand keeping these dangerous capabilities out of the fuse process is attractive for security reasons.  Maybe it could be delegated to the watchdog process so we don't need yet another process.

I thought about keeping a cache of recent cvmfs-info strings based on PID (not including the `%{path}` substitution because that changes) but wasn't sure about the implications for locking.  So I decided to leave that for a later optimization if needed.